### PR TITLE
Partition symbols

### PR DESCRIPTION
**Pull request by kolesnikovae**
This is a major change of how profile symbols are handled and stored in Pyroscope.

We previously introduced **`symdb`**, a specialized database that handles and stores stack traces. Along with that, we've introduced the concept of a partition which is supposed to delimit symbolic information of independent applications and thereby improve data locality and query selectivity. However, only the stack traces are actually partitioned. Locations, functions, mappings, and strings are stored in large unordered parquet tables. This requires us to load all of these tables into memory at query time because rows are accessed randomly in most cases. To make matters worse, these tables are never paged out of memory.

The PR contains three major changes:
 - Parquet tables are no longer loaded entirely into memory. `locations`, `functions`, `mappings`, and `strings` parquet tables are now partitioned in the same way as stack traces, and the symdb index stores references to the partition row ranges. On query, relevant row ranges are read from the block directly.
 - Everything related to symbols refactored to symdb (`symbols` block directory). Some parts are taken as is (e.g dedup slice), some were rewritten completely.

The changes result in a very substantial decrease of memory consumption of ingesters and storage-gateways (see more [in the issue](https://github.com/grafana/pyroscope/issues/2025)). In the screenshot below you can see the trend of memory utilization of storage-gateways of one of the test clusters:

image

Downsides:
 1.  In some cases, the query latency may be slightly higher. Fetching symbols from blocks has a nearly constant latency of a few hundred milliseconds and is done asynchronously with the read of samples, so if there are very few profiles matching the query, the overall query execution duration is generally higher than it used to be.
 2. Block size might be potentially higher because if the same application is being profiled under different names, its symbolic information is stored multiple times. In practice, however, the increase might be moderate (~10%) or unnoticeable.

Future work:
 * Improve partitioning heuristics.
 * Optimize write path.
 * Optimize tree implementation (unordered insert, in-memory truncation, arena-based allocator).
 * **`symdb`** caching: most frequently used partitions should be kept in memory (locally).
 * Standalone symbolizer service.

Resolves:
 - https://github.com/grafana/pyroscope/issues/2025
 - https://github.com/grafana/pyroscope/issues/2061

### CODE REVIEW

**Review by simonswine — COMMENTED** [review]
I would say I have not fully digested it just yet. 

The good news is this code passes the correctness test from the hackathon, but that bad one is `next` also does.

Let's sync tomorrow about running a test in our dev environment 🙂

File: pkg/phlaredb/block/metadata.go [hunk 1]
```diff
@@ -39,6 +39,7 @@ const (
 	// Version1 is a enumeration of Phlare section of TSDB meta supported by Phlare.
 	MetaVersion1 = MetaVersion(1)
 	MetaVersion2 = MetaVersion(2)
+	MetaVersion3 = MetaVersion(3)
```
[the remark this answers is not in the document]
Comment (kolesnikovae) [reply pkg/phlaredb/block/metadata.go]: I'm going to properly document the changes in the schema design at some point in the future once it is stable. For now I added version notes: https://github.com/grafana/pyroscope/pull/2196/commits/213ff6b09f03d422386f482c1f8af714b79dfd9e

**Review by petethepig — APPROVED** [review]
We're trying to put this in v1, so I'm approving it so that we can deploy it to dev and ops sooner.

File: pkg/phlaredb/block_querier.go [hunk 2]
```diff
@@ -1221,151 +1079,3 @@ func repeatedColumnIter[T any](ctx context.Context, source Source, columnName st
 	opentracing.SpanFromContext(ctx).SetTag("columnName", columnName)
 	return query.NewRepeatedPageIterator(ctx, rows, source.RowGroups(), column.ColumnIndex, 1e4)
 }
-
-type ResultWithRowNum[M any] struct {
```
Comment (cyriltovena) [inline pkg/phlaredb/block_querier.go:1225]: :feelsgood:

**Pull request merged**